Add support for impersonation tokens

maxenced opened the pull request:
Impersonation token are only returned once, as response to the create request.

That's why we use `Set` inside the  `create` method instead of read.

Tested on latest gitlab-ce

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,136 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			"user": {
```
Other resources in this provider use `user_id`. Can you change it?

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,136 @@
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"expires_at": {
```
Can you add a `ValidateFunc` to this value?

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,136 @@
+	// We need to set token here instead of read, as it is only returned once
+	d.Set("token", impersonationToken.Token)
+
+	return resourceGitlabUserImpersonationTokenRead(d, meta)
```
I'm curious why you need to call the read function if you already have the data you need from the `impersonationToken` object?

armsnyder replied:
Never mind. I see it is a recommended practice in the docs. Carry on!

maxenced replied:
guess I can resolve this ?

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,136 @@
+func resourceGitlabUserImpersonationTokenDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	usertoken := strings.Split(d.Id(), "/")
+	userId, err := strconv.Atoi(usertoken[0])
```
Clean up these ineffectual err assigns please. If you don't use it, make it `_`. 😉 

I also saw this in your `testAccGitlabUserImpersonationTokenDestroy` method, so please check over the code.

maxenced replied:
removed in both resource and test (as much as I can see)

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,136 @@
+// so the object still exists on Gitlab side, but we remove it from TF state
+func resourceGitlabUserImpersonationTokenDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	usertoken := strings.Split(d.Id(), "/")
```
Looks like other resources in this provider use `:` as the ID delimiter. Can you do this also?

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token_test.go:
```diff
@@ -0,0 +1,239 @@
+
+		token, resp, err := conn.Users.GetImpersonationToken(userId, tokenId, nil)
+		if err == nil {
+			if token != nil && fmt.Sprintf("%d", token.ID) == rs.Primary.ID && token.Revoked == false {
```
```suggestion
			if token != nil && fmt.Sprintf("%d", token.ID) == rs.Primary.ID && !token.Revoked {
```

looks cleaner

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token_test.go:
```diff
@@ -0,0 +1,239 @@
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not Found: %s", n)
```
nit pick but errors should not be capitalized

maxenced replied:
Removed the couple of capitalized "F"

armsnyder commented:
Heyya @maxenced are you still working on this PR? I left some comments for you. 😄

maxenced commented:
> Heyya @maxenced are you still working on this PR? I left some comments for you. smile

Oh yes, sorry for having missed (or likely, forgotten) your comments. Fixed them all (as far as I can tell) and rebased.

armsnyder commented on website/docs/r/user_impersonation_token.html.markdown:
````diff
@@ -0,0 +1,56 @@
+
+## Example Usage
+
+```hcl
````
Can you fix the indentation in the example to be consistent?

armsnyder commented on website/docs/r/user_impersonation_token.html.markdown:
```diff
@@ -0,0 +1,56 @@
+## Importing user impersonation token
+
+You can import an impersonation token to terraform state using `terraform import <resource> <id>`.
+The `id` must be formated like `<user_id>/<token_id>`, for example:
```
The format is actually `<user_id>:<token_id>` 😉

armsnyder reviewed (changes requested):
Be sure also to update the `website/gitlab.erb` docs index file please.

maxenced commented:
> Be sure also to update the `website/gitlab.erb` docs index file please.

fixed with last push

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token_test.go:
```diff
@@ -0,0 +1,239 @@
+		tokenId, _ := strconv.Atoi(usertoken[1])
+
+		token, resp, err := conn.Users.GetImpersonationToken(userId, tokenId, nil)
+		if err == nil {
+			if token != nil && fmt.Sprintf("%d", token.ID) == rs.Primary.ID && !token.Revoked {
+				return fmt.Errorf("Impersonation token still exists ")
+			}
+		}
+		if resp.StatusCode != 404 {
+			return err
+		}
```
I'm having trouble understanding the `if resp.StatusCode != 404` bit here. What was your intent? Would this work instead?

```suggestion
		if err != nil {
			return err
		}
		if !token.Revoked {
			return errors.New("impersonation token still exists")
		}
```

armsnyder commented on gitlab/resource_gitlab_user_impersonation_token_test.go:
```diff
@@ -0,0 +1,239 @@
+	ExpiresAt *gitlab.ISOTime
+}
+
+func testAccCheckGitlabUserImpersonationTokenAttributes(n string, token *gitlab.ImpersonationToken, want *testAccCheckGitlabUserImpersonationTokenExpectedAttributes) resource.TestCheckFunc {
```
Can you also check `scopes`? And check that `created_at` is non-zero?

armsnyder commented on website/docs/r/user_impersonation_token.html.markdown:
```diff
@@ -0,0 +1,56 @@
+You can import an impersonation token to terraform state using `terraform import <resource> <id>`.
+The `id` must be formated like `<user_id>:<token_id>`, for example:
+
+    terraform import gitlab_user_impersonation_token.example 21/25
```
```suggestion
    terraform import gitlab_user_impersonation_token.example 21:25
```

armsnyder commented:
> > Be sure also to update the `website/gitlab.erb` docs index file please.
> 
> fixed with last push

Hmm I don't see it in the diff. Maybe it got lost during rebase?

timofurrer commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,137 @@
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
```
Please upgrade to the [terraform SDK v2](https://www.terraform.io/plugin/sdkv2/sdkv2-intro) and make sure to rebase to an up-to-date `master` branch.

timofurrer commented on gitlab/resource_gitlab_user_impersonation_token.go:
```diff
@@ -0,0 +1,137 @@
+
+func resourceGitlabUserImpersonationToken() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabUserImpersonationTokenCreate,
```
Please use the context-aware CRUD functions, see the [upgrade guide](https://www.terraform.io/plugin/sdkv2/guides/v2-upgrade-guide#more-support-for-context-context).

Also make sure to use the new `ctx` to pass it to the `gitlab` API functions as `gitlab.WithContext(ctx)`. It's usually the last parameter of the API functions.

timofurrer reviewed (changes requested):
Please rebase to latest master as we've upgraded to the SDK v2. There are also some open comments from @armsnyder and myself. Please check them out - otherwise - lgtm :)

timofurrer commented:
@maxenced do you still want to work on this? I think it would make a good candidate for the next release!
